src/DependencyInjection/DI/src/Resources.resx

```diff
@@ -156,4 +156,7 @@
   <data name="ImplementationTypeCantBeConvertedToServiceType" xml:space="preserve">
     <value>Implementation type '{0}' can't be converted to service type '{1}'</value>
   </data>
+  <data name="AsyncDisposableServiceDispose" xml:space="preserve">
+    <value>'{0}' type only implements IAsyncDisposable. Use DisposeAsync to dispose the container.</value>
```

Comment:
I'm not 100% sure this is reasonable. There might be cases where something is added to the container out of somebody's control and it'll cause hosting to just crash and burn when the container gets disposed.

Reply:
They can always DisposeAsync().Wait() to workaround

Reply:
Hosting is the one calling Dispose on the container.

Reply:
Maybe make it an option so people can work around it in cases where they don't own the container (like Azure Functions)? By default we can throw but you can turn it off to get the sync over async behavior back.

Reply:
Both us and Azure Functions are going to move to DisposeAsync() in time for 3.0 and I doubt that there are a lot of types that implement only IAsyncDisposable

Reply:
I can add a flag if you insist.

Reply:
Our templates are calling Dispose not DisposeAsync, it's likely this will be the case until we ship the RTM of .NET Core 3.0. How do we mitigate this in the mean time?

Reply:
Why RTM? I'll make a change to hosting after merging this PR. IAsyncDisposable is netcoreapp3.0 only feature so backcompat is not a concern here.

Reply:
Because the templates need a higher c# version to use async main, in the templates. Today we call Run. This will also be breaking if somebody updates from 2.0 to 3.0 and components start adding IAsyncDisposable things. Their templates will have the old call to Run and nothing will tell people they need to change that to be an async main with RunAsync

Reply:
I don't understand, `.Run` already calls `RunAsync` and blocks on it so templates would work just fine.

https://github.com/aspnet/Hosting/blob/master/src/Microsoft.AspNetCore.Hosting/WebHostExtensions.cs#L65

Reply:
Ah true, so I guess this error would never happen unless you explicitly called Dispose on the host.

Reply:
Maybe TestHost scenarios have the potential to break?

Reply:
Nope, WebHost.Dispose() blocks StopAsync and `provider.DisposeAsync` calls would get moved there 
https://github.com/aspnet/AspNetCore/blob/master/src/Hosting/Hosting/src/Internal/WebHost.cs#L350

Reply:
You shouldn't call Dispose in StopAsync. Thats not right. Stopping needs to happen before disposing, not during.

Reply:
You a right, Dispose will call DisposeAsync and block.